src/escrow.rs

```diff
@@ -0,0 +1,231 @@
+pub trait EscrowInternal {
+    /// Identifier over which the escrow exists
+    type Id: BorshSerialize;
+    /// State stored inside the lock
+    type State: BorshSerialize + BorshDeserialize;
+
+    /// Retrieve the state root
+    fn root() -> Slot<()> {
+        Slot::root(DefaultStorageKey::Escrow)
+    }
+
+    /// Inner function to retrieve the slot keyed by it's `Self::Id`
+    fn locked_slot(&self, id: &Self::Id) -> Slot<Self::State> {
+        Self::root().field(StorageKey::Locked(id))
+    }
+
+    /// Read the state from the slot
+    fn get_locked(&self, id: &Self::Id) -> Option<Self::State> {
+        self.locked_slot(id).read()
+    }
+
+    /// Set the state at `id` to `locked`
+    fn set_locked(&mut self, id: &Self::Id, locked: &Self::State) {
+        self.locked_slot(id).write(locked);
+    }
+
+    /// Clear the state at `id`
+    fn set_unlocked(&mut self, id: &Self::Id) {
+        self.locked_slot(id).remove();
+    }
+}
+
+/// Some escrowable capabilities, with a simple locking/unlocking mechanism
+/// If you add additional `Approve` capabilities here, you can make use of a step-wise locking system.
+pub trait Escrow {
+    /// Identifier over which the escrow exists
+    type Id: BorshSerialize;
+    /// State stored inside the lock
+    type State: BorshSerialize + BorshDeserialize;
+
+    /// Lock some `Self::State` by it's `Self::Id` within the store
+    fn lock(&mut self, id: &Self::Id, state: &Self::State);
+
+    /// Unlock and release some `Self::State` by it's `Self::Id`
+    ///
+    /// Optionally, you can provide a handler which would allow you to inject logic if you should unlock or not
+    fn unlock(&mut self, id: &Self::Id, lock_handler: impl FnOnce(&Self::State) -> bool);
+
+    /// Check if the item is locked
+    fn is_locked(&self, id: &Self::Id) -> bool;
+}
+
+impl<T> Escrow for T
+where
+    T: EscrowInternal,
+    <T as EscrowInternal>::Id: Serialize,
+    <T as EscrowInternal>::State: Serialize,
+{
+    type Id = <Self as EscrowInternal>::Id;
+    type State = <Self as EscrowInternal>::State;
+
+    fn lock(&mut self, id: &Self::Id, state: &Self::State) {
+        require!(self.get_locked(id).is_none(), ESCROW_ALREADY_LOCKED_MESSAGE);
+
+        self.set_locked(id, state);
+        Lock {
+            id: id.to_owned(),
+            locked: Some(state),
+        }
+        .emit();
+    }
+
+    fn unlock(&mut self, id: &Self::Id, lock_handler: impl FnOnce(&Self::State) -> bool) {
+        let lock = self
+            .get_locked(id)
+            .unwrap_or_else(|| panic_str(ESCROW_NOT_LOCKED_MESSAGE));
+
+        if lock_handler(&lock) {
+            self.set_unlocked(id);
+            Lock::<_, Self::State> { id, locked: None }.emit();
+        } else {
+            panic_str(ESCROW_LOCK_HANDLER_FAILED_MESSAGE)
+        }
+    }
+
+    fn is_locked(&self, id: &Self::Id) -> bool {
+        self.get_locked(id).is_some()
+    }
```

Comment:
This function doesn't require `Self::Id: Serialize` or `Self::State: Serialize`. Thoughts on moving those bounds to the trait definition, or removing the events? Compare with `Rbac`, which also doesn't expose a public interface, and doesn't have events.